base/collection_xattr_common.go

```diff
@@ -204,21 +239,33 @@ func (c *Collection) WriteUpdateWithXattrs(ctx context.Context, k string, xattrK
 		// Attempt to write the updated document to the bucket.  Mark body for deletion if previous body was non-empty
 		deleteBody := value != nil
 		if updatedDoc.IsTombstone {
-			casOut, writeErr = c.WriteTombstoneWithXattrs(ctx, k, exp, cas, updatedDoc.Xattrs, deleteBody, opts)
+			casOut, writeErr = c.WriteTombstoneWithXattrs(ctx, k, exp, cas, updatedDoc.Xattrs, updatedDoc.XattrsToDelete, deleteBody, opts)
 		} else {
-			casOut, writeErr = c.WriteWithXattrs(ctx, k, exp, cas, updatedDoc.Doc, updatedDoc.Xattrs, opts)
+			if wasTombstone {
+				if len(updatedDoc.XattrsToDelete) > 0 {
+					return 0, sgbucket.ErrDeleteXattrOnTombstone
+				}
+				casOut, writeErr = c.WriteResurrectionWithXattrs(ctx, k, exp, updatedDoc.Doc, updatedDoc.Xattrs, opts)
+			} else {
+				casOut, writeErr = c.WriteWithXattrs(ctx, k, exp, cas, updatedDoc.Doc, updatedDoc.Xattrs, updatedDoc.XattrsToDelete, opts)
+			}
 		}
 
 		if writeErr == nil {
 			return casOut, nil
 		}
 
-		if IsCasMismatch(writeErr) {
+		if previousLoopCas != nil && *previousLoopCas == cas {
+			err := RedactErrorf("Stopping an infinite loop trying to update doc with xattr for key=%s, xattrKeys=%s", UD(k), UD(xattrKeys))
```

Comment:
I think this error would be better written as something like "CAS retry triggered, but no change in document CAS detected"